apps/doji-backend/src/review/review.service.ts

```diff
@@ -0,0 +1,61 @@
+import { Review } from '@backend/entities/Review'
+import { Session } from '@backend/entities/Session'
+import { ReviewAverageRatingDTO, ReviewCreationRequestDTO } from '@backend/review/review.dto'
+import { IUserReference } from '@libs/api'
+import { EntityRepository, UniqueConstraintViolationException } from '@mikro-orm/core'
+import { InjectRepository } from '@mikro-orm/nestjs'
+import { Injectable, NotFoundException, UnprocessableEntityException } from '@nestjs/common'
+
+@Injectable()
+export class ReviewService {
+  constructor(
+    @InjectRepository(Review) private readonly reviewRepo: EntityRepository<Review>,
+    @InjectRepository(Session) private readonly sessionRepo: EntityRepository<Session>,
+  ) {}
+
+  async createReview(dto: ReviewCreationRequestDTO, userRef: IUserReference) {
+    const user = await userRef.getUser()
+    const session = await this.sessionRepo.findOne({ id: dto.sessionId })
+    if (!session) {
+      throw new NotFoundException('Session not found')
+    }
+    const newReview = new Review()
+    newReview.user = user
+    newReview.session = session
+    newReview.content = dto.content
+    newReview.rating = dto.rating
+    try {
+      await this.reviewRepo.persistAndFlush(newReview)
+    } catch (e) {
+      if (e instanceof UniqueConstraintViolationException) {
+        throw new UnprocessableEntityException('User with this username or email already exist.')
+      } else {
+        throw e
+      }
+    }
+    return
+  }
+
+  async getAllReview(sessionId: number) {
+    const session = await this.sessionRepo.findOne({ id: sessionId })
+    if (!session) {
+      throw new NotFoundException('Session not found')
+    }
+    const reviewList = await this.reviewRepo.find({ session: session })
+    return reviewList
+  }
+
+  async getReviewAverageRatingById(sessionId: number) {
+    const session = await this.sessionRepo.findOne({ id: sessionId })
+    if (!session) {
+      throw new NotFoundException('Session not found')
+    }
+    const reviewList = await this.reviewRepo.find({ session: session })
+    const count = reviewList.length
+    let sum = 0
+    reviewList.forEach((value) => (sum += value.rating))
+    const avgRating = new ReviewAverageRatingDTO()
+    avgRating.avgRating = sum / count
+    return avgRating
```

Comment:
ตรงนี้ถ้าเรา query แล้วมาคำนวณ average จะเสียเวลา เราสามารถใช้ query builder ในการเขียนคำสั่ง SQL ได้  (คำนวณ avg ใน db)https://stackoverflow.com/questions/63096550/mikro-orm-how-to-get-results-from-queries-with-aggregates-into-their-own-entity

Reply:
แล้วก็อาจริง ๆ เราส่งไปมันไปสร้างกับ `getAllReviews` เลยก็ได้ ไม่ต้องส่งแยก

Reply:
ก็คือจะให้ ใช้ query builder คำนวณ avg แล้วส่งไปพร้อมกับ getAllReview เลยใช่ไหม แล้วก็ไม่ต้องทำ reduce แล้ว

Reply:
See [this snippet](https://github.com/2110336-SoftwareEngineering2-2021-2/sec2_ChocoTech/pull/72/files#diff-ed260796fb09c487d5885a9dd22cb3ec82700b66a42c4705f2f75168ed47af1fR97-R123)